### PR TITLE
Improved error message for calling/constructing types

### LINKED ISSUE #32013: Terrible error message for calling/constructing types
https://twitter.com/sebmck/status/1141776682275049472

image

# Suggestions

```
This expression is not callable.
  Type '{0}' has no call signatures.
```

For unions:

```
This expression is not callable.
  No constituent of type '{0}' is callable.
```

```
This expression is not callable.
  Not all constituents of type '{0}' are callable.
    Type '{0}' has no call signatures.
```

```
This expression is not callable.
  Each member of the union type '{0}' has signatures, but none of those signatures are compatible with each other.
```

The same would be done for constructors.

```
This expression is not constructable.
  Type '{0}' has no construct signatures.
```

```
This expression is not constructable.
  No constituent of type '{0}' is constructable.
```

```
This expression is not constructable.
  Not all constituents of type '{0}' are constructable.
    Type '{0}' has no construct signatures.
```

```
This expression is not constructable.
  Each member of the union type '{0}' has signatures, but none of those signatures are compatible with each other.
```

### PR DESCRIPTION
**Pull request by dragomirtitian**
Fixes #32013

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21987,11 +21987,83 @@ namespace ts {
...
+                    Diagnostics.This_expression_is_not_callable :
+                    Diagnostics.This_expression_is_not_constructable
+            );
+        }
         function invocationError(node: Node, apparentType: Type, kind: SignatureKind, relatedInformation?: DiagnosticRelatedInformation) {
-            const diagnostic = error(node, (kind === SignatureKind.Call ?
-                Diagnostics.Cannot_invoke_an_expression_whose_type_lacks_a_call_signature_Type_0_has_no_compatible_call_signatures :
-                Diagnostics.Cannot_use_new_with_an_expression_whose_type_lacks_a_call_or_construct_signature
-            ), typeToString(apparentType));
+            const diagnostic: Diagnostic = createDiagnosticForNodeFromMessageChain(node, invocationErrorDetails(apparentType, kind));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: What's with the annotation?

**Comment by DanielRosenwasser** [comment]
I really like this improvement, but I'm starting to get uneasy about the wording.

- "This expression" really refers to the target of the call, but we're highlighting the whole range.
- We can change the range to focus on the invoked expression, but I think we still *want* the current range.
- The wording is so nice and simple! I don't want to change that!

**Comment by DanielRosenwasser** [comment]
In other words, can we find better wording that refers to the invoked expression without sounding terrible?

File: src/compiler/checker.ts (hunk 1 above)
Comment (dragomirtitian) [reply src/compiler/checker.ts]: Mistake, will delete

**Comment by dragomirtitian** [comment]
@DanielRosenwasser Personally I think the range of this error would be better on the target. I think it would make the most sense, the problem is with the target of the call, the arguments have no bearing on the error and are just extra noise. Unlike for example when overload selection fails when it makes sense to highlight the whole call expression since the problem is with the combination of call target and arguments. 

I think changing the target might cause some problems with some quick fixes (`fixInvalidImportSyntax` uses the node for this specific error), but nothing that can't be fixed 😊

**Comment by DanielRosenwasser** [comment]
Yeah, I think you have a good point. I would try to change the range to the target if possible. Maybe see #31414 for the better range. If you'd be up for fixing the ranges on overload errors to include arguments, that'd be awesome too.

**Comment by dragomirtitian** [comment]
@DanielRosenwasser Sure I will look into it. Not sure about a time line, this week is a bit busy though.

**Comment by dragomirtitian** [comment]
@DanielRosenwasser 

What do programmers do Friday night ? They code for fun 😁

I had a chance to finish the changes we discussed above to error spans. The behavior will be as follows:

1. If the error is related to the call target (the call target is not callable), the span only includes the target (not the arguments).
```ts
declare const test: string;
test();
~~~~
```
1. a. If the call target is part of a property access expression, only the last name is part of the span.
```ts
declare const o: { test: string }
o.test();
  ~~~~
```
2. If the error is related to overload resolution, the whole call expression (including arguments) is part of the error span
```ts
declare const test: (o:string) => void
test();
~~~~~~
```
a. If the call target is part of a property access expression, only the last name is part of the span (together with the arguments).
```ts
declare const o: { test(o:string): void }
o.test();
  ~~~~~~
```
I think this is a nice improvement over #31414 and also improves error spans on long invocations chains where the offending call has a non-callable target (something #31414 did not do). 

I know @sandersn is also working in this area (#32092) hope this will not conflict with that too much.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Pull request merged**